pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,243 @@
+
+		host:      host,
+		namespace: namespace,
+		logger:    log.DefaultLogger(),
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+		n.logger.Infof("Node-labeller cannot work without KVM device.")
+		return
+	}
+
+	n.configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			n.Queue.Add(key)
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			n.Queue.Add(key)
+		},
+		UpdateFunc: func(_, new interface{}) {
+			key, _ := cache.MetaNamespaceKeyFunc(new)
+			n.Queue.Add(key)
+		},
+	})
+
+	go n.configMapInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.configMapInformer.HasSynced)
+
+	n.nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
+			node := new.(*v1.Node)
+			n.Queue.Add(node.Name)
+		},
+	})
+
+	go n.nodeInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.nodeInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+
+	err := n.run()
+
+	if err != nil {
+		n.Queue.AddRateLimited(key)
+	} else {
+		n.Queue.Forget(key)
+	}
+	return true
+}
+
+func (n *NodeLabeller) run() error {
+	defer n.lock.Unlock()
+	cpuFeatures := make(map[string]bool)
+	cpuModels := make([]string, 0)
+
+	//parse all informations
+	cpuModels, cpuFeatures, err := n.getCPUInfo()
+	if err != nil {
+		n.logger.Infof("node-labeller cannot get new labels %s\n", err.Error())
+		return err
+	}
+
+	//load node
+	nodeObj, exists, err := n.nodeInformer.GetStore().GetByKey(n.host)
+	if err != nil || !exists {
+		n.logger.Infof("node-labeller cannot get node %s", n.host)
+		return err
+	}
+	var (
+		ok   bool
+		node *v1.Node
+	)
+	if node, ok = nodeObj.(*v1.Node); !ok {
+		n.logger.Infof("node-labeller cannot convert node " + n.host)
+		return fmt.Errorf("Could not convert node " + n.host)
+	}
+
+	n.lock.Lock()
+	originalNode := node.DeepCopy()
+
+	//prepare new labels
+	newLabels := n.prepareLabels(cpuModels, cpuFeatures)
+	//remove old labeller labels
+	n.removeCPULabels(node, n.getNodeLabellerLabels(node))
+	//add new labels
+	n.addNodeLabels(node, newLabels)
+	//patch node only if there is change in labels
+	if !reflect.DeepEqual(node.Labels, originalNode.Labels) {
+		//patch node with new labels
+		data := []byte(fmt.Sprintf(`[{"op": "replace", "path": "/metadata/labels", "value": {%s}}, {"op": "replace", "path": "/metadata/annotations", "value": {%s}}]`, convertMapToText(node.Labels), convertMapToText(node.Annotations)))
```

Comment:
ah, i think i forgot to mention this earlier.

There's a trick we need to do here to ensure we don't accidently stomp on a label change that our cached node object might not have yet.

Check out what we're doing here https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/watch/vmi.go#L393

```
			test := fmt.Sprintf(`{ "op": "test", "path": "/status/conditions", "value": %s }`, string(oldConditions))
			patch := fmt.Sprintf(`{ "op": "replace", "path": "/status/conditions", "value": %s }`, string(newConditions))
			log.Log.V(3).Object(vmi).Infof("Patching VMI conditions")
			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(fmt.Sprintf("[ %s, %s ]", test, patch)))
```

That `test` operator contains the expected "previous" value. If the test operation fails, then the replace won't occur. You can use this technique with the labels to prevent accidentally dropping a label that might have been added at the same moment the Patch is being executed.